### ISSUE TITLE
Should pointerover and pointerenter work on mobile?

### ISSUE THREAD

**Issue by trusktr** [w3c/pointerevents#346]
I can't seem to get them to work on mobile in Chrome (on Android, or in the desktop simulator in devtools).

This demo works on desktop: 

https://codepen.io/trusktr/pen/898505353c12d3001b74ba0e2a0e5c72 (EDIT: it doesn't fully work. It has the issues described in the [below comment](https://github.com/w3c/pointerevents/issues/346#issuecomment-759211255))

(click on the grid cell that the cube is on, and drag it to other cells, to move the cube).

However, on mobile, the `pointerdown` event works, but the `pointerover` events on the cells never fire, so we can not drag the cube's placement on mobile. 

I tried both `pointerover` and `pointerleave` but no luck.

Is it per spec? Or a browser bug?

**Comment by patrickhlauke** [comment]
You'll likely need to also add `touch-action` somewhere, as by default arbitrary `pointerover`/`pointerleave` will not be fired simply when a finger is moved over an area. https://w3c.github.io/pointerevents/#declaring-candidate-regions-for-default-touch-behaviors

**Comment by NavidZ** [comment]
What Patrick mentioned is correct for most of the cases we see developers report. Actually that is what I suspected at first as well. But in your particular case there is another reason as you indeed keep getting the pointermoves so there is no pointercancel going on. So as far as I tested the touch-action none would not solve the issue for you.

You are seeing what is specified as an [implicit pointer capture](https://w3c.github.io/pointerevents/#implicit-pointer-capture) for touch pointers in the spec which makes what you see working as intended. Adding this code to your sample page removes the capture and makes the behavior the way you expect. Basically I'm removing the capture as soon as I see one on the cells you work with.

`cells.forEach(n => n.addEventListener('gotpointercapture', e => e.target.releasePointerCapture(e.pointerId)));`

But I need to mention a **big caveat** this solution has. Basically you are asking the browser to continue hit-testing for you for all incoming pointer events and that does make things generally slower (but maybe not as much as you notice in your testing). So something to be aware of and actually saving this hit-testing time was the reason we introduced the implicit pointer capture at first.

**Comment by patrickhlauke** [comment]
ah, good catch @NavidZ - i should have checked more thoroughly

**Comment by trusktr** [comment]
Wow. Thanks. That was not intuitive.

The document says it is in the name of performance. Are there benchmarks?

**Comment by NavidZ** [comment]
There is no external benchmark that I'm aware of. The cost of the hit-testing certainly depends on your page structure and most certainly different browsers are doing it different. I remember EdgeHTML was claiming there were quite fast but well they switched to Blink these days.

**Comment by trusktr** [comment]
Is this a premature optimization that is only making things confusing (i.e. people expect pointerover to happen like the name implies, but by default it doesn't)?

I have never experienced any performance issues with `mouseover`, so I assume `pointerover` would perform just as fast.

**Comment by NavidZ** [comment]
It was years ago and obviously the numbers might have changed recently. Off the top of my head I remember the hit-testing cost considering the high sampling rate of touch screens on phones were high enough (we do have internal metrics on that) for us not wanting to introduce that regression. Note that it would have been a regression vs the state of the web at the time that `touchmove` events don't do hit-test and always go to the first element you touch. To translate that world to the `pointerevent` world we had to introduce this notion of implicit capture.

To put the numbers into perspective if you would like to achieve a 60fps page then you would have ~16 ms time budget for each frame. Obviously different parts of this budget goes towards different parts of the browser doing the work as well as developer doing some js logic. But having devices that do 120Hz sampling of touch moves on phones (meaning you might be getting two touch moves in a frame) spending 1-2 ms on hit-testing would eat quite a lot of the budget in a frame.

Also note that we are running things on a mobile device for touch screens mostly and the processing power + battery consideration vs a desktop environment that you would usually expect mouse events are different. So that was why web at the time didn't care doing hit-testing on all mouse move events and most likely why you didn't experience any performance issues with `mouseover` stuff. It was mostly desktop and it could handle it. Obviously over time as mouse devices also started to up the sampling to like 1000Hz we started introducing other features such as [coaleased events](https://w3c.github.io/pointerevents/#dom-pointerevent-getcoalescedevents) to reduce the overhead of hit-testing and whatnot but that's just some boring history :)

**Comment by trusktr** [comment]
> To translate that world to the pointerevent world we had to introduce this notion of implicit capture.

Ah, I see. Thanks for explaining!

> I remember the hit-testing cost considering the high sampling rate of touch screens on phones were high enough
> But having devices that do 120Hz sampling of touch moves on phones (meaning you might be getting two touch moves in a frame)

I believe there is no need to run the events any faster than the animation loops (requestAnimationFrame) can run. From what I've seen in Chrome, user events fire at most once between animation frames (in the testing I've done). Which makes sense: no need to calculate things more than once just to render it once. If rAF is 60fps, then events should be at most 60fps too.

`getCoalescedEvents` is interesting (and probably hardly used, I've never heard of it and I've been around for a while). I see that it could be useful for calculating a better movement curve for example, but the handler still doesn't need to run more than once per rAF. No handler ever needs to run faster than the screen will re-draw, I believe.

And these days phones are fast. I play http://krunker.io on my phone.

Would applications break if the default was changed to no capturing by default? (Also the terminology can be confused with event capture phases).

**Comment by trusktr** [comment]
Seems like we need it only for performance, but we already have performance options. For example, we can load a page with

```css
body {
  pointer-events: none;
}

.selectively-enable-pointer-events-for-certain-elements {
  pointer-events: auto;
}
```

and control which elements will be hit tested. Seems like it would lead to a more intuitive developer experience, _and_ people can optimize as needed.

I'm willing to bet that the vast majority of the time, hit testing won't be an issue (even if everything is hit tested).

**Comment by patrickhlauke** [comment]
CSS `pointer-events` has nothing to do with JS Pointer Events.

**Comment by trusktr** [comment]
Seems like it does, because `pointer-events: none` disables JS pointer events, and  `pointer-events: auto` enables them.

Mozilla [docs say](https://developer.mozilla.org/en-US/docs/Web/CSS/pointer-events)

> The `pointer-events` [CSS](https://developer.mozilla.org/en-US/docs/Web/CSS) property sets under what circumstances (if any) a particular graphic element can become the [target](https://developer.mozilla.org/en-US/docs/Web/API/Event/target) of pointer events.

and `target` links to the docs on `Event.target`, which is a JavaScript API.

**Comment by NavidZ** [comment]
> getCoalescedEvents is interesting (and probably hardly used, I've never heard of it and I've been around for a while). I see that it could be useful for calculating a better movement curve for example, but the handler still doesn't need to run more than once per rAF. No handler ever needs to run faster than the screen will re-draw, I believe.

Before we introduced getCoalescedEvents we had to actually send all the events (and maybe multiple in a frame) to the js (as say a drawing app wanted to see all the events to be able to draw a smooth(er) curve) and we couldn't just limit it to once per frame as calling the js event handler was the only way to deliver all the events back then and hence the hit-testing for all of them. But getCoalescedEvents API (that will be used for the drawing apps and others) allowed us to actually limit the hit-testing and dispatching one event per animation loop that you see today in Chrome.

> Seems like it does, because pointer-events: none disables JS pointer events, and pointer-events: auto enables them.

The part that I don't get regarding your proposal about `pointer-events` is the performance benefits. It's been quite some time that I looked at the Chromium code and particularly the hit-testing but "pointer-events: none" on parts of your page doesn't necessarily increase the performance of hit-testing. At the end users will touch the parts of the page that are interactive and you want them to interact with. So again `pointer-events: none` on some non-interactive elements (that users aren't going to touch anyway) isn't going to make things faster. If anything it might make it slower as we have to add layers for those regions and some special cases.

Anyhow, this is just all history. APIs can come and go and indeed change (although hardly due to compat issues). If there are enough interests in a particular subject or maybe the ergonomics of an API I'm sure other people will be onboard as well. I have to say I don't have much more to add here in terms of the future direction as I no longer actively work in Chrome. But feel free to file another issue outlining your proposal so others can chime in. This issue doesn't seem particularly relevant to what you have in mind regarding the ergonomics.

**Comment by trusktr** [comment]
Is there any hit testing at all with capturing?

**Comment by NavidZ** [comment]
Theoretically no. When capture is in place the target of the event stream that is captured is always the capturing element and user agents can skip hit-testing.

That being said I remember in Chromium implementation we used to do the hit-test for mouse pointers even when are captured. I don't recall whether we fixed it or not.

**Comment by trusktr** [comment]
Seems that:

- if `pointer-events` is configured such that one element has `pointer-events: auto` (all others are disabled because `html {pointer-events: none}` is in place), 
  - If the element has no children, there should be zero hit testing.
  - If the element has any children, then only those children are hit tested (makes sense).

So, for an element with class `.container` that has no children, then

```css
html { pointer-events: none; }
.container { pointer-events: auto; }
```

would be the same performance as the capture that we described above, and for an element with children, we can selectively include which things should be hit tested. Either,

```css
html { pointer-events: none; }
.container { pointer-events: auto; }
```

to hit-test all children and grand-children of `.container`, or


```css
html { pointer-events: none; }
.container .specific-elements { pointer-events: auto; }
```

to hit-test only particular elements that are (grand)children of `.container`.

It would be more intuitive for this to be the case, with `pointerover`, `pointerout`, `pointerenter`, and `pointerleave` working by default, without the capture idea in the picture. I guess I haven't has a use case that ran into the OP issue yet (or I did things a different way last times), despite how long I've been in web development (probably not as long as you great folks).

In fact, that's how I always expected `pointer-events` to work. That is intuitive.

---

In the renowned [Qt](http://qt.io/) framework, there are no events enabled for any surfaces, by default. One must specify all event areas, which is akin to my previous `html {pointer-events: none}` idea paired with selective enabling of areas that should be hit tested.

**Comment by trusktr** [comment]
I updated the pen in the OP, and that one works fine on mobile now.

But for the life of me, I just can't get it to work in the following fork of the demo, in which I want to start the `pointerdown` event on the `cube-` element instead of on the grid cells. 

A `gotpointercapture` event never fires on the `lume-node`s (grid cells), or on the `.gridContainer` (MDN says the event bubbles).

https://codepen.io/trusktr/pen/2d143d1f00c35c0bf6a82c9017b70806

But also, the `pointerup` event on `document` doesn't always fire after each drag, which I don't understand.

So I'm not sure if I'm still having the capture problem, or if I'm doing something silly (if I am, it isn't obvious), or if there's a bug.

To reproduce, try dragging the cube around to different cells multiple times (let go, and start again), and it will fail sometimes. When it fails, the cube won't move while dragging, or it will move only once, but once the pointer is lifted (and `pointerup` event failed to fire) then the cube continues to move around.

EDIT: I take it back, the original demo also exhibits the same issue for me now that I tried more: https://codepen.io/trusktr/pen/eYdPOKb (this is the original demo, but with the a call to `releasePointerCapture` added)

**Comment by trusktr** [comment]
Looks like browsers are buggy with Pointer Events still, or the specs don't specify enough for them. The problems I described do not happen for me in Firefox, but only in Chrome.

Firefox on the other hand has a different issue where it is impossible to drag the cube to the top-right cell of the grid. Other than that, no issues dragging the cube over and over.

**Comment by trusktr** [comment]
I updated the original demo so it works perfectly (in Chrome) by not using `pointerover` and instead coupling `pointermove` events to the size of the grid and grid cells:

Fully working (drag works perfectly every time, Chrome):

https://codepen.io/trusktr/pen/eYdPOKb

Broken (drag fails sometimes, try multiple times until it fails):

https://codepen.io/trusktr/pen/OJRBQWX

After much fiddling, I believe there are two browser bugs:

- https://bugs.chromium.org/p/chromium/issues/detail?id=1166044
- https://bugzilla.mozilla.org/show_bug.cgi?id=1686390

If they aren't bugs, then I missed something, in which case it would be proof that the Pointer Events API is too complicated and not intuitive.

**Comment by trusktr** [comment]
^ Can you help me determine why the broken example doesn't work, or if there is in fact browser bugs?

**Comment by antoine10245** [comment]
I can't believe this is closed...I just spent 2 days trying to do a simple pointermove event. all opened tickets are merged or close and nothing been solved. It seems it's coming from the fact that event.pointerId got incremented each time the pointer hit boundaries of a container.
as @trusktr suggested using:
`event.target.releasePointerCapture(event.pointerId);`
at the right time will avoid the the pointerId to be incremented. only then onpointermove will work as expected. thx @trusktr
